rusk-wallet/src/lib/clients.rs

```diff
@@ -245,15 +256,27 @@ impl StateClient for State {
         prompt::status("Notes received!");
 
         // collect notes
-        let notes: Vec<Note> = res
+        let mut fresh_notes: Vec<Note> = res
             .into_iter()
-            .map(|n| {
+            .flat_map(|n| {
                 let mut bytes = [0u8; Note::SIZE];
                 bytes.copy_from_slice(&n);
-                Note::from_bytes(&bytes).unwrap()
+
+                let note = Note::from_bytes(&bytes).unwrap();
+                let key = note.hash().to_bytes().to_vec();
+                match cached_notes.contains_key(&key) {
+                    true => None,
+                    false => Some(note),
+                }
             })
             .collect();
-        Ok(notes)
+
+        cache::persist_notes(&self.cache, psk, &fresh_notes[..]);
+        cache::persist_blockheight(&self.cache, psk, actual_block);
```

Comment:
Once again, feels like it would make more sense to have the cache own the `Connection` and the `StateClient` own the initialized `Cache`. We avoid calling these static methods and passing the connection object.